cou/steps/analyze.py

```diff
@@ -0,0 +1,221 @@
+    status: ApplicationStatus
+    config: dict
+    model_name: str
+    charm: str = ""
+    charm_origin: str = ""
+    os_origin: str = ""
+    channel: str = ""
+    units: defaultdict[str, dict] = field(default_factory=lambda: defaultdict(dict))
+
+    def __post_init__(self) -> None:
+        """Initialize the Application dataclass."""
+        self.charm = extract_charm_name_from_url(self.status.charm)
+        self.channel = self.status.charm_channel
+        self.charm_origin = self.status.charm.split(":")[0]
+        self.os_origin = self._get_os_origin()
+        for unit in self.status.units.keys():
+            workload_version = self.status.units[unit].workload_version
+            self.units[unit]["workload_version"] = workload_version
+            os_version = self._get_current_os_version(workload_version)
+            self.units[unit]["os_version"] = os_version
+
+    def __hash__(self) -> int:
+        """Hash magic method for Application.
+
+        :return: Unique hash identifier for Application object.
+        :rtype: int
+        """
+        return hash(f"{self.name}{self.charm}")
+
+    def __eq__(self, other: Any) -> bool:
+        """Equal magic method for Application.
+
+        :param other: Application object to compare.
+        :type other: Any
+        :return: True if equal False if different.
+        :rtype: bool
+        """
+        return other.name == self.name and other.charm == self.charm
+
+    def __str__(self) -> str:
+        """Dump as string.
+
+        :return: Summary representation of an Application.
+        :rtype: str
+        """
+        summary = {
+            self.name: {
+                "model_name": self.model_name,
+                "charm": self.charm,
+                "charm_origin": self.charm_origin,
+                "os_origin": self.os_origin,
+                "channel": self.channel,
+                "units": {
+                    unit: {
+                        "workload_version": details.get("workload_version", ""),
+                        "os_version": details.get("os_version", ""),
+                    }
+                    for unit, details in self.units.items()
+                },
+            }
+        }
+        yaml = YAML()
+        with StringIO() as stream:
+            yaml.dump(summary, stream)
+            return stream.getvalue()
+
+    def _get_representative_workload_pkg(self) -> str:
+        """Get the representative package name of a charm workload.
+
+        :return: Package name that represents the charm workload. E.g: cinder-common
+        :rtype: str
+        """
+        try:
+            package = CHARM_TYPES[self.charm]["representative_workload_pkg"]
+        except KeyError:
+            logging.warning(
+                "Representative workload package not found for application: %s", self.name
+            )
+            package = ""
+        return package
+
+    def _get_current_os_version(self, workload_version: str) -> str:
```

Comment:
@gabrielcocenza is working on in another branch. I think it would we better if we postpone this.

Reply:
I'm also not very happy using the external `get_os_code_info`. This function showed that it's not able to handle cases that the major version it's not enough to determine the openstack release (E.g: charm-openstack-dashboard) or when we can have multiple openstack releases options (Ceph, rabbitmq and etc). Moreover, i think it would be cleaner using [packaging](https://packaging.pypa.io/en/latest/version.html) instead of regex.

My current idea to improve this is to have a lookup like this:

```
python3
from collections import OrderedDict, namedtuple
Versions = namedtuple("Versions", "min_version max_version")

openstack_lookup = {
    "keystone": OrderedDict(
        [
            ("ussuri", Versions(Version("17.0.0"), Version("17.0.1"))),
            ("victoria", Versions(Version("18.0.0"), Version("18.1.0"))),
        ]
    )
}
```
All the `min_version` and `max_version` would be mapped from the [openstack-release](https://releases.openstack.org/victoria/index.html) table.

Reply:
I'll file a bug for tracking this separately